auction-server/src/simulator/main.rs

```diff
@@ -0,0 +1,191 @@
+            .await?;
+        let accounts = accounts_with_context.value;
+        let mut accounts_config = AccountsConfig::new();
+        let mut programs_to_fetch = vec![];
+        for (account_key, account) in keys.iter().zip(accounts.iter()) {
+            if let Some(account) = account {
+                if let Ok(UpgradeableLoaderState::Program {
+                    programdata_address,
+                }) = account.state()
+                {
+                    programs_to_fetch.push((*account_key, programdata_address));
+                } else {
+                    accounts_config
+                        .accounts
+                        .insert(*account_key, account.clone());
+                }
+                // TODO: handle lookup tables
+            } else {
+                // it's ok to not have an account (this account is created by the transaction)
+            }
+        }
+
+        //TODO: handle caching for programs already fetched
+
+        let program_accounts = self
+            .receiver
+            .get_multiple_accounts_with_commitment(
+                &programs_to_fetch
+                    .iter()
+                    .map(|(_, programdata_address)| *programdata_address)
+                    .collect::<Vec<_>>(),
+                CommitmentConfig::processed(),
+            )
+            .await?
+            .value;
+        for ((program_key, _), program_account) in
+            programs_to_fetch.iter().zip(program_accounts.iter())
+        {
+            if let Some(program_account) = program_account {
+                accounts_config
+                    .programs
+                    .insert(*program_key, program_account.clone());
+            } else {
+                // it's not ok to point to a non-existent program. TODO: handle this case
+            }
+        }
+        Ok(Response {
+            value:   accounts_config,
+            context: accounts_with_context.context,
+        })
+    }
+
+    pub async fn simulate_transaction(
+        &self,
+        transaction: &VersionedTransaction,
+    ) -> RpcResult<Result<SimulatedTransactionInfo, FailedTransactionMetadata>> {
+        let accounts_config_with_context = self.fetch_tx_accounts(&[transaction.clone()]).await?;
+        let accounts_config = accounts_config_with_context.value;
+        let pending_txs = self.fetch_pending_and_remove_old_txs().await;
+        let mut svm = LiteSVM::new()
+            .with_sigverify(false)
+            .with_blockhash_check(false)
+            .with_transaction_history(0);
+        accounts_config.apply(&mut svm);
+
+        pending_txs.into_iter().for_each(|tx| {
+            let _ = svm.send_transaction(tx);
+        });
+        let res = svm.simulate_transaction(transaction.clone());
+        Ok(Response {
+            value:   res,
+            context: accounts_config_with_context.context,
+        })
+    }
+}
```

Comment:
one way i was thinking of structuring this is the following:

- you have an rpc client that informs your view of state of all accounts except for a list of accounts which txs that you have submitted have altered (i.e. writable accounts in txs submitted by searchers)
- for each of those writable accounts, you pull the state of the account when you submit the tx and log that as well as the state of the account post simulating your tx
- you regularly poll the state of the accounts in your list as well as listen for commitment updates on the transactions you submitted. as soon as you observe either a change in the RPC's view of an account state or get a commitment update on a transaction, you drop your simulated view of the account from your list
- when simulating transactions, you check if any of the transaction's accounts are in your list. if so, you don't fetch those accounts' data and instead use your stored accounts; you fetch all other accounts.
- this allows an easy fallback where if the original transaction you submitted fails or doesn't reach `processed` within enough time, then you revert to on-chain state. but for a short period of time after you submit, you will definitely retain a future view of state so you can simulate new txs more intelligently

Reply:
of course, the key is in how you poll and balance btwn keeping ahead of future tx inclusion and fallback behavior in case a transaction doesn't reach `processed` or fails on-chain. if another tx changes the state of an account that your submitted tx touches before it is `processed` you should probably pessimistically drop your simulated forward state bc your tx may not succeed any longer. practically speaking, for limo orders, this shouldn't happen too often, but it's worth thinking about the tradeoffs here in case of other accounts that users may interact with outside of this context